File: Dynamic-Programming/337. House Robber III/houseRobberIII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var rob = function (root) {
    return Math.max(...dfs(root));
    function dfs(root) {
        if (!root) return [0, 0];

        let left = dfs(root.left);
        let right = dfs(root.right);

        let withRoot = root.val + left[1] + right[1];
        let withoutRoot = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);

        return [withRoot, withoutRoot];
    }
};

File: Dynamic-Programming/337. House Robber III/houseRobberIII.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: TreeNode) -> int:
        def dfs(root):
            if root is None:
                return [0, 0]
            
            left = dfs(root.left)
            right = dfs(root.right)
            
            withRoot = root.val + left[1] + right[1]
            withoutRoot = max(left[0], left[1]) + max(right[0], right[1])
            
            return [withRoot, withoutRoot]
        
        return max(dfs(root))